### PR TITLE
Introduce-machine

### PR DESCRIPTION
**Pull request by arnolddevos**
Introduce Machine and KvStore

### CODE REVIEW

File: event-driven/src/lib.rs [hunk 1]
```diff
@@ -80,9 +82,46 @@ pub trait Fsm {
...
+
+    /// remove and return accumulated messages.
+    fn drain_all(&mut self) -> impl Iterator<Item = Self::Item> + Send;
+}
+
 #[cfg(test)]
-mod tests {
-    use super::*;
+mod test {
+    use super::{Change, Fsm, Input};
```
Comment (huntc) [inline event-driven/src/lib.rs:124]: Why not `use super::*;` as it was? It is generally what you want.

File: kv_store/src/async_query.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+fn respond_many<F, V, R>(func: F, sender: oneshot::Sender<R>) -> RespondMany<V>
+where
+    F: FnOnce(&dyn Iterator<Item = (&Path, &V)>) -> R + Send + 'static,
+    R: Send + 'static,
+{
+    Box::new(|vs| {
+        let _ = sender.send(func(vs));
+    })
+}
```
Comment (huntc) [inline kv_store/src/async_query.rs:104]: It'd be nice to see some tests around this.

File: kv_store/src/path.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        self.items.is_empty()
+    }
+}
+
+/// One element of a `Path` can be a number or a name.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug, From)]
+pub enum PathItem {
+    Number(u64),
+    Name(String),
+}
```
Comment (huntc) [inline kv_store/src/path.rs:51]: Putting this in as a reminder to consider smol_str.

File: Cargo.toml [hunk 4]
```diff
@@ -4,5 +4,23 @@ resolver = "2"
...
     "event-driven",
-    "event-driven-macros"
+    "event-driven-macros",
+    "kv_store",
+    "machine",
+    "streambed-machine",
 ]
+
+[workspace.package]
+description = "Event Driven Finite State Machine library"
```
Comment (huntc) [inline Cargo.toml:14]: Does the description belong at this level?

File: Cargo.toml [hunk 5]
```diff
@@ -4,5 +4,23 @@ resolver = "2"
...
+readme = "README.md"
+version = "0.7.0"
+edition = "2021"
+license = "Apache-2.0"
+repository = "https://github.com/titanclass/edfsm.git"
```
Comment (huntc) [inline Cargo.toml:19]: `rust-version = "1.70.0"` ?

File: Cargo.toml [hunk 6]
```diff
@@ -4,5 +4,23 @@ resolver = "2"
...
+
+[workspace.dependencies]
+edfsm = { path = "event-driven", version = "0.7.0" }
+event-driven-macros = { path = "event-driven-macros", version = "0.7.0" }
+streambed-machine = { path = "streambed-machine", version = "0.1.0" }
+machine = { path = "machine", version = "0.7.0" }
+kv_store = { path = "kv_store", version = "0.7.0" }
```
Comment (huntc) [inline Cargo.toml:26]: We'll also need any path based dependencies to go in here along with the version that they are to be published at. For example, in streambed:

```
streambed = { path = "streambed", version = "0.11.1" }
streambed-confidant = { path = "streambed-confidant", version = "0.11.1" }
streambed-logged = { path = "streambed-logged", version = "0.11.1" }
streambed-test = { path = "streambed-test", version = "0.11.1" }
```

File: kv_store/README.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# KV Store
+
+A KV Store based on edfsm FSM.
```
Comment (huntc) [inline kv_store/README.md:3]: Just as a TODO - these READMEs will need to be beefed up of course.

File: streambed-machine/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    /// Send one item to the underlying commit log.
+    pub async fn produce(&self, item: A) -> Result<Offset, ProducerError> {
+        let key = item.record_key();
+        let topic = self.topic.clone();
+
+        if let Some(value) = self.codec.encode(item).await {
+            self.commit_log
+                .produce(ProducerRecord {
+                    topic,
+                    headers: Vec::new(),
```
Comment (huntc) [inline streambed-machine/src/lib.rs:71]: Would `vec!([])` be more efficient?

File: streambed-machine/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    }
+}
+
+impl<L> CommitLogExt for L where L: CommitLog {}
+
+impl<L, C, A> LogAdapter<L, C, A>
+where
+    L: CommitLog,
+    C: Codec<A>,
+    A: RecordKey + Clone + 'static,
```
Comment (huntc) [inline streambed-machine/src/lib.rs:60]: `RecordKey` should no longer be required given that we can default key to 0.

File: streambed-machine/src/lib.rs (hunk 8 above)
Comment (arnolddevos) [reply streambed-machine/src/lib.rs:71]: Nah, check the docs: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.new "The vector will not allocate until elements are pushed onto it."

File: event-driven/src/lib.rs (hunk 1 above)
Comment (arnolddevos) [reply event-driven/src/lib.rs:124]: use with * is an anti-pattern IMO. albeit, super::* is not as bad as all that.

File: streambed-machine/src/lib.rs (hunk 9 above)
Comment (arnolddevos) [reply streambed-machine/src/lib.rs]: OK

File: kv_store/src/async_query.rs (hunk 2 above)
Comment (arnolddevos) [reply kv_store/src/async_query.rs:102]: Well, I copied that line from a canonical source... https://github.com/streambed/streambed-rs/blob/330fb377e8526c71bf72b5d695cfda121771baf6/streambed-patterns/src/ask.rs#L47

Comment (huntc) [reply kv_store/src/async_query.rs:102]: Hah. I actually didn't write the original in streambed. ;-)

File: kv_store/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,36 @@
+[package]
+name = "kv_store"
```
Comment (huntc) [inline kv_store/Cargo.toml:2]: We may need to rename `ed-kv-store` due to publishing constraints (there is already a kv-store`. Note also the use of `-` and not `_`.

I'm also happy to rename the `event-driven` and `event-driven-macros` dirs using their respective Cargo names.

File: kv_store/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+description = "A KV Store based on edfsm FSM"
```
Comment (huntc) [inline kv_store/Cargo.toml:3]: "on the event-drive FSM"...

File: kv_store/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+version.workspace = true
+edition.workspace = true
+license.workspace = true
+repository.workspace = true
+
+[dependencies]
+edfsm.workspace = true
+machine = { workspace = true, optional = true }
+smol_str = { version = "0.3.2", default-features = false, features = ["serde"] }
```
Comment (huntc) [inline kv_store/Cargo.toml:14]: `version = "0.3.2"` can be replaced with `workspace = true` with `smol_str` declared there.

File: kv_store/Cargo.toml [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+derive_more = { version = "1", features = [
+    "from",
+    "try_into",
+    "deref",
+    "into_iterator",
+] }
+tokio = { version = "1", features = ["sync"], optional = true }
+embassy-sync = { version = "0.6.0", optional = true }
+serde = { version = "1", features = ["derive"] }
```
Comment (huntc) [inline kv_store/Cargo.toml:23]: Workspace dependencies to be used.

File: kv_store/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+[dev-dependencies]
+machine.workspace = true
+tokio = { version = "1", features = ["full"] }
+serde = "1"
+serde_json = "1"
+serde_qs = "0.13"
```
Comment (huntc) [inline kv_store/Cargo.toml:30]: Workspace dependencies.

File: kv_store/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+    /// Get all the entries
+    GetAll(RespondMany<V, ()>),
+
+    /// Get the value at the given path or None and emit an event for that path.
+    Upsert(Path, RespondOne<V, E>),
+
+    /// Get all the entries and emit an event for a particular (usually new) path.
+    Insert(RespondMany<V, Keyed<E>>),
+}
```
Comment (huntc) [inline kv_store/src/lib.rs:68]: Is it useful that these types remain public given the `Requester`?

File: machine/src/error.rs [hunk 16]
```diff
@@ -0,0 +1,45 @@
+use derive_more::From;
+
+/// Result type for this module
+pub type Result<A> = std::result::Result<A, Error>;
```
Comment (huntc) [inline machine/src/error.rs:4]: Std? or core...

Comment (huntc) [reply machine/src/error.rs:4]: May be also re-check for other occurrences of std.

File: streambed-machine/README.md [hunk 17]
```diff
@@ -0,0 +1,3 @@
+# Streambed Simplified
+
+Simplified interface to streambed CommitLog
```
Comment (huntc) [inline streambed-machine/README.md:3]: Hmm. That implies streambed is complex. ;-) How about "Streambed integration".

Also needs beefing up.

**Comment by huntc** [comment]
Looking really, really nice.

General comment: perhaps no longer a draft (will require squash and merge, or squashing prior).

...and perhaps some mentions in the repo's README about these new crates.

File: machine/Cargo.toml [hunk 18]
```diff
@@ -0,0 +1,30 @@
+[package]
+name = "machine"
```
Comment (huntc) [inline machine/Cargo.toml:2]: `ed-machine`

File: streambed-machine/Cargo.toml [hunk 19]
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "streambed-machine"
```
Comment (huntc) [inline streambed-machine/Cargo.toml:2]: I think we can move `streambed-machine` to streambed once the other crates are published.

File: machine/Cargo.toml (hunk 18 above)
Comment (huntc) [reply machine/Cargo.toml:2]: I was wondering if we might reconsider the `Machine` type name also given that the FSM is already the "machine"... `EventDriver` or something?

File: machine/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
+/// A `Machine` is a state machine (implementing `Fsm`) running in a rust `task`.
+///
+/// Each `Machine` has a current state, an input channel, an effector and adapters.
+/// The type of the state and the channel are part of the state machine specification.
+/// As messages arrive on the channel, the state machine `Fsm::step` method
+/// evolves the state and orchestrates side effects.
+///
+/// The effector enables the machine to execute side effects and its type is also part of
+/// the state machine specification, `Fsm::SE`.  Generally, side effects must
+/// be synchronous and if they may block they should be bracketed with tokio's `block_in_place`.
```
Comment (huntc) [inline machine/src/lib.rs:48]: I think this para needs updating given the use of the output buffer.

File: machine/src/output.rs [hunk 21]
```diff
@@ -0,0 +1,32 @@
+use edfsm::{Drain, Init};
+use std::vec::Vec;
+
+#[derive(Debug)]
+pub struct OutputBuffer<A>(pub Vec<A>);
```
Comment (huntc) [inline machine/src/output.rs:5]: I think we should also have a noop type that implements drain and init for those situations where there is no output on an FSM. This is the case for our KV store types where we are generally interested in just the events emitted, and don't need the output.

File: kv_store/src/lib.rs (hunk 15 above)
Comment (arnolddevos) [reply kv_store/src/lib.rs:68]: Yes very useful.  Requester is feature gated for tokio.   kv_store is allocated but no_std.

**Comment by arnolddevos** [comment]
Incorporated into #19

**Pull request closed without merging**